wp-includes/class-wp-offline-page.php

```diff
@@ -0,0 +1,157 @@
+<?php
+/**
+ * WP_Offline_Page class.
+ *
+ * @package PWA
+ */
+
+/**
+ * WP_Offline_Page class.
+ * Much of this is taken from wp-admin/privacy.php.
+ */
+class WP_Offline_Page {
+
+	/**
+	 * The option group for the offline page option.
+	 *
+	 * @var string
+	 */
+	const OPTION_GROUP = 'reading';
+
+	/**
+	 * The option name.
+	 *
+	 * @var string
+	 */
+	const OPTION_NAME = 'wp_offline_page';
+
+	/**
+	 * The ID of the settings section.
+	 *
+	 * @var string
+	 */
+	const SETTING_ID = 'pwa_offline_page';
+
+	/**
+	 * Inits the class.
+	 */
+	public function init() {
+		add_action( 'admin_init', array( $this, 'register_setting' ) );
+		add_action( 'admin_init', array( $this, 'settings_field' ) );
+	}
+
+	/**
+	 * Registers the offline page setting.
+	 */
+	public function register_setting() {
+		register_setting(
+			self::OPTION_GROUP,
+			self::OPTION_NAME,
+			array(
+				'type'              => 'integer',
+				'sanitize_callback' => array( $this, 'sanitize_callback' ),
+			)
+		);
+	}
+
+	/**
+	 * Sanitize callback for the setting.
+	 * Mainly taken from wp-admin/privacy.php.
+	 *
+	 * @todo: Ensure this is stored with a custom post status.
+	 * @param string $raw_setting The setting before sanitizing it.
+	 * @return string|null The sanitized setting, or null if it's invalid.
+	 */
+	public function sanitize_callback( $raw_setting ) {
+		$sanitized_post_id = sanitize_text_field( $raw_setting );
+		$offline_page      = get_post( $sanitized_post_id );
+		if ( ! $offline_page instanceof WP_Post ) {
+			add_settings_error(
+				self::OPTION_NAME,
+				self::OPTION_NAME,
+				__( 'The current offline page does not exist. Please select or create one.', 'pwa' )
+			);
+		} elseif ( 'trash' === $offline_page->post_status ) {
+			add_settings_error(
+				self::OPTION_NAME,
+				self::OPTION_NAME,
+				__( 'The current offline page is in the trash. Please select or create one.', 'pwa' )
+			);
+		} else {
+			return $sanitized_post_id;
+		}
+	}
+
+	/**
+	 * Adds a settings field to the 'Reading Settings' page.
+	 */
+	public function settings_field() {
+		add_settings_field(
+			self::SETTING_ID,
+			__( 'Progressive Web App Offline Page', 'pwa' ),
```

Comment:
Instead of “Progressive Web App Offline Page” I think this should be something like:

>  Page displays when offline

This would align more with the existing:

> Your homepage displays

Then there can be a `p.description` element which explains that this is is for PWA, and how provide more details about how it works, maybe mentioning it is “like a 404 page but it shows when someone's internet is down (e.g. during flight or going through tunnel), as opposed to when content is not found.”

Reply:
@westonruter Is this what you are thinking:

![pwa-wp-pr28](https://user-images.githubusercontent.com/7284611/42601018-dcc80472-8529-11e8-9eba-73314d96073d.png)